### ISSUE TITLE
OIPC - Automated Notifications and Comments

### ISSUE THREAD

**Issue by m-prodan** [bcgov/foi-flow#4694]
* As a an assignee to a request
* I want to be notified when a request is flagged for OIPC review
* so that I know action may need to be taken

**Assumptions & Scope**	
This story covers the various notification and automated comment scenarios that  will flow from the OIPC stories.

What is IN scope?
Notifications and Comments when an OIPC review is created/added
Notifications and Comments when an OIPC review is closed
Notifications and Comments when an OIPC review has a pending comply date

What is NOT in scope?

**Acceptance Criteria**	

**Scenario 1: OIPC Complaint/Review Created**
* GIVEN an IAO user has created an OIPC complaint/review on a file
* WHEN they save the details
* THEN all assigned users except the user that entered the complaint/review will get a notification saying: "OIPC [review type] opened for [reason]"
* AND this message will be posted in the request history comments

**Scenario 2: OIPC Review/Complaint Closed**
* GIVEN an IAO user is on the OIPC details page
* WHEN they make a selection on the outcome
* THEN all assigned users, except the user that entered the outcome, will get a notification saying: "OIPC [review type] closed for [Outcome]"
* AND this message will be posted in the request history comments

**Scenario 3: OIPC Review/Complaint - Inquiry Outcome**
* GIVEN an IAO user is on the OIPC details page
* WHEN they make a selection on the Inquiry outcome
* THEN all assigned users, except the user that entered the outcome, will get a notification saying: "OIPC [review type] Inquiry Decision: [Outcome]"
* AND this message will be posted in the request history comments

**Scenario 4: Compliance Date - Notification**
* GIVEN an IAO user is on the OIPC details page
* WHEN they enter the Inquiry compliance date and order no
* THEN all assigned users, except the user that entered the inquiry information, will get a notification saying: "OIPC Inquiry Order [Order no.] compliance date due [Date]"
* AND this message will be posted in the request history comments

**Scenario 5: Compliance Date - Reminder Notification - 2 days**
* GIVEN an IAO user has entered a compliance date
* AND the OIPC review is still open
* WHEN they it is 2 business days before the compliance date
* THEN all assigned users, will get a notification saying: "OIPC Inquiry Order [Order no.] compliance date due [Date]"
* AND this message will be posted in the request history comments

**Scenario 6: Compliance Date - Reminder Notification - 1 day**
* GIVEN an IAO user has entered a compliance date
* AND the OIPC review is still open
* WHEN they it is 1 business day before the compliance date
* THEN all assigned users, will get a notification saying: "OIPC Inquiry Order [Order no.] compliance date due [Date]"
* AND this message will be posted in the request history comments

**Scenario 6: Compliance Date - Reminder Notification - Due Date**
* GIVEN an IAO user has entered a compliance date
* AND the OIPC review is still open
* WHEN they it is the day of the compliance date
* THEN all assigned users, will get a notification saying: "OIPC Inquiry Order [Order no.] compliance due today. 
* AND this message will be posted in the request history comments

**Dependencies? What is the impact of this dependency? (If so, link dependency in the ticket, make it visible in a team´s backlog)**

**Validation Rules? (If yes, list here)**	

**Design**
@xxx - please link the Design here

**Definition of Ready**

1. [ ] Is there a well articulated User Story?
2. [ ] Is there Acceptance Criteria that covers all scenarios (happy/sad paths)?
3. [ ] If there is a user interface, is there a design? 
4. [ ] Does the user story need user research/validation?
5. [ ] Does this User Story needs stakeholder approval?
6. [ ] Design / Solution accepted by Product Owner
7. [ ] Is this user story small enough to be completed in a Sprint? Should it be split?
8. [ ] Are the dependencies known/ understood? (technical, business, regulatory/policy)
9. [ ] Has the story been estimated?

**Definition of Done**

1.    [ ] Passes developer unit tests
2.    [ ] Passes peer code review
3.    [ ] If there's a user interface, passes UX assurance
4.    [ ] Passes QA of Acceptance Criteria with verification in Dev and Test
5.	[ ] Confirm Test cases built and succeeding 
6.	[ ] No regression test failures
7.	[ ] Test coverage acceptable by Product Owner
8.	[ ] Ticket ready to be merged to master or story branch
9.	[ ] Developer to list Config changes/ Update documents and designs
10.	[ ] Can be demoed in Sprint Review 
11.	[ ] Tagged as part of a Release
12.   [ ] Feature flagged if required
13.  [ ] Change Management activities done?

**Comment by m-prodan** [comment]
Notifications/Comments story for OIPC review.

cc @JHarrietha-AOT @lmullane @liseandtea @KyEggleston

**Comment by JHarrietha-AOT** [comment]
No designs created for notifications and comments - should follow all others currently in system

**Comment by divyav-aot** [comment]
do we need a comment/notification for deletion? - **yes, we need a comment, no notification**

assigned users meaning? is it including watchers? - **Including the AssingedTo, MinistryAssginedTo and Watchers**

Complaint due date  - **all users should get it. Assigned users, created user and watchers**

**Other Notifications, it is already mentioned**

Once it is complied with, how we understand it is done? - **Once all the complaint has the outcome, no need to send a notification. And if the request is closed and all outcomes are selected then no need to show in the queue. The flag should be on**

**Comment by sumathi-thirumani-aot** [comment]
[Private Zenhub Image](https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBeEFmQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--b65afd561f0f5725008249fc56e2c77bf0c4d551/image.png)

Information missing on "OIPC No" in comment & notification.

**Comment by sumathi-thirumani-aot** [comment]
@JHarrietha-AOT For scenario#2; the message format has to be "OIPC [review type] closed for [**reason**]" .; whereas currently this is given as ""OIPC [review type] closed for [Outcome]" -> This is like "OIPC [review type] closed for Closed".

**Comment by nimya-aot** [comment]
@sumathi-thirumani-aot Message is not appearing on the 'Request History comments'  when an OIPC review is open. could you please check?
<img width="1522" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBd0pOQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--2304d9ce45670ef0a524a64e2fd6e4b2eb886e80/image.png" alt="image.png" />

**Comment by nimya-aot** [comment]
@sumathi-thirumani-aot scenario 3 also failing. When we select an inquiry outcome, there is no message on 'Request History comments'

**Comment by nimya-aot** [comment]
@sumathi-thirumani-aot If we create  OIPC complaint today and comply date is today itself, Then also assignee not getting notification . Could you please check?
<img width="1499" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBenhOQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--6fe4be90a32ba2ef962237d0c32c89502ba46d8d/image.png" alt="image.png" />

**Comment by sumathi-thirumani-aot** [comment]
Works as expected. The cron job to send notificaiton runs only once before business start. This is existing logic..

**Comment by nimya-aot** [comment]
@sumathi-thirumani-aot Sorry. But the messages is not showing up on request history, that issue resolved?

**Comment by sumathi-thirumani-aot** [comment]
Changes done on the comment type to display under system generated.
<img width="1848" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMmRXQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--616cfa92088f1acb12140b4596074679d7ae7b84/image.png" alt="image.png" />

**Comment by sumathi-thirumani-aot** [comment]
PR for review (dev-rook) -> https://github.com/bcgov/foi-flow/pull/4883

**Comment by nimya-aot** [comment]
@liseandtea do we need notification for an already closed OIPC?
<img width="1446" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBMnBuQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--5681045f0c3c1866db0cfb255037b908fce50e8e/image.png" alt="image.png" />

**Comment by nimya-aot** [comment]
@sumathi-thirumani-aot scenario 3 is not working when a user creates an OIPC complaint first time and also enters the inquiry outcome same time. Could you please check this request? Notification/ message on request history comments not showing up for the scenario 3.
https://test-rook-foirequest.apps.silver.devops.gov.bc.ca/foi/foirequests/23/ministryrequest/23

**Comment by nimya-aot** [comment]
>@sumathi-thirumani-aot scenario 3 is not working when a user creates an OIPC complaint and enters the inquiry outcome same time. Could you please check this request? Notification/ message on request history comments not showing up for the scenario 3 on that case. 
>May be needs to confirm with business at stand up.
>>
@liseandtea confirmed it is fine with only one reminder notification if user creates a OIPC request and enters inquiry outcome at the same time. No need of separate notifications as in  scenario 3

**Comment by KyEggleston** [comment]
Do we have a story/AC for notifications when a request is closed, but has an active OIPC Review? 

cc: @liseandtea @lmullane @JHarrietha-AOT

**Comment by nimya-aot** [comment]
@KyEggleston currently we have story #4679 for the closed request having active OIPC review. But notifications are not mentioned in that. May be we can add here that

**Comment by sumathi-thirumani-aot** [comment]
Regarding the reminder notifications. I suppose we don't need to send notification for any outcome( This should not be only for Closed). Please confirm. Right now exclusion is only for closed<img width="2000" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM2QvQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--9a70eaad3d735f26439979c5cc11abcf3bc26ba8/image.png" alt="image.png" />

**Comment by sumathi-thirumani-aot** [comment]
Got confirmation from Jacky and Lise; we dont need to send reminder notification; if any outcome is chosen

**Comment by sumathi-thirumani-aot** [comment]
Unit Testing screenshot
<img width="1103" src="https://api.zenhub.com/attachedFiles/eyJfcmFpbHMiOnsibWVzc2FnZSI6IkJBaHBBM3gvQWc9PSIsImV4cCI6bnVsbCwicHVyIjoiYmxvYl9pZCJ9fQ==--f46c42820b0c9e5591c9a0e7a0e07be416f3ebfe/image.png" alt="image.png" />

**Comment by nimya-aot** [comment]
working as expected, so closing
